rfcs/20190116-embedding-partitioned-variable.md

```diff
@@ -0,0 +1,498 @@
+|:-------- |:-------- |:----------------|
+|Mirrored/ CollectiveAllReduce Strategy|Pre-TF 2.0: <br>w/ Keras' Embedding layer API: won't be sharded but will be mirrored on each replica. <br><br>Post-TF 2.0: <br>w/ Distribution Strategy's API directly: allow sharding; being placed on host; or dense update. <br><br>Heuristics in Embedding layer to auto-set performance improvement hints.|Pre-TF 2.0: <br>w/ Keras' layer API: won't be sharded but will be mirrored on all replicas, just like a normal layer. <br><br>Post-TF 2.0: w/ Distribution Strategy's API directly: allow being mirrored on each host. <br><br>Heuristics in Embedding layer to auto-place larger embeddings on host.|
+|ParameterServer Strategy|Pre-TF 2.0:<br>w/ Keras' Embedding layer API: probably sharded, according to `partitioner` passed to its constructor.<br><br>Post-TF 2.0:<br>w/ Distribution Strategy's API directly: allow creating partitioned variable with different partitioners.|Pre-TF 2.0:<br>w/ Keras' layer API: probably sharded, according to `partitioner` passed to its constructor which applies globally to all layers.<br><br>Post-TF 2.0:<br>w/ Distribution Strategy's API directly: allow creating partitioned variable with different partitioners.|
+
+In the `ParameterServerStrategy`, the goal of supporting partitioned variable is to match the existing behavior in 1.x. We will propose its APIs for TF 2.0 below. 
+
+**In the mirrored and collective allreduce architecture, we will only mirror embeddings or logits layers, either on devices or on host.** 
+
+In the future, we will probably support a new strategy that is a hybrid of these two architectures in the future.
+
+Please refer to the discussion in [Alternatives Considered](#Alternatives-Considered) section for why we've made these decisions.
+
+
+### Proposed Change for pre-TF 2.0
+
+
+The following is planned to be done before TF 2.0's release.
+
+
+#### A new type `ShardedVariable`
+
+
+As discussed above, `PartitionedVariable` sometimes concatenates component variables silently. So we need another object as a convenience wrapper for a list of variables and use this as a guard against any inadvertent concatenation. Users don't have to deal with it as long as they use Keras' layer API and it won't be exposed as a public API.
+
+In order not to be mixed up with the 1.x API, we give it a different name (other options are `LoadBalanceVariable` or `VariableList`): 
+
+```python
+class ShardedVariable(object):
+  """Wraps a list of variables."""
+  def __init__(self, name, shape, dtype, variable_list,
+               partitions):
+    pass
+  
+  def get_variable_list(self):
+    """Returns a list of partition variables."""
+    pass
+```
+
+Users may have to concatenate variables themselves when dealing with it directly in post-TF 2.0 and using it for loadbalancing purposes rather than embeddings. More convenience methods will be added post-TF 2.0.
+
+
+#### Distribution strategy
+
+
+##### ParameterServerStrategy
```

Comment:
There should be a way to create PartitionedVariables still for the load-balancing case, right?

I don't see it here. Clarify?

Reply:
Done. Layers should create load-balancing variables if `partitioner` is specified in `ParameterServerStrategy`.